// nframe/lib/global.js
var util = require('util');
var res;
var req;
var $

var pre = {
	_GET : {},
	_POST : {},
	_COOKIE : {},
	_REQUEST : {}
}

exports.set = function(request, response) {
	res = response;
	req = request;
	init();
};

init = function() {
	initURL();
	initGET();
	initPOST();
	initCOOKIE();
	initREQUEST();
	initSESSION();
}

initURL = function() {
	return req.url.slice(0, req.url.indexOf('?'));
};

initGET = function() {
	pre._GET = {};
	if(req.url.indexOf('?') !== -1) {
		var urlparts = req.url.slice(req.url.indexOf('?') + 1);
		var query = urlparts.split('&');
		for (var p=0; p < query.length; ++p) {
			var pair = query[p].split('=');
			if(pair[0] != undefined || pair[1] != undefined) {
				pre._GET[pair[0]] = pair[1];
			}
		}
	}
};

initPOST = function() {
	pre._POST = {};
	var body = '';
	req.on('data', function(chunk) {
		body += chunk;
		if (body.length > 1e6) {
			req.connection.destroy();
		}
	});
	req.on('end', function() {
		var pairs = body.split('&');
		for (var p=0; p < pairs.length; ++p) {
			var pair = pairs[p].split('=');
			pre._POST[pair[0]] = pair[1];
		}
	});
};

initCOOKIE = function() {
	pre._COOKIE = {};
	if (req.headers.cookie) {
		var cookies = req.headers.cookie.split(';');
		for (var c=0; c < cookies.length; ++c) {
			var pair = cookies[c].split('=');
			pre._COOKIE[pair[0]] = pair[1];
		}
	}
};

initREQUEST = function() {
	pre._REQUEST = {};
	if (pre._GET) {
		for (var k in pre._GET) {
			pre._REQUEST[k] = pre._GET[k];
		}
	}
	if (pre._POST) {
		for (var k in pre._POST) {
			pre._REQUEST[k] = pre._POST[k];
		}
	}
	if (pre._COOKIE) {
		for (var k in pre._COOKIE) {
			pre._REQUEST[k] = pre._COOKIE[k];
		}
	}
};

/** All the sessions of all the users. */
var sessions = {};
initSESSION = function() {
	if ((typeof pre._COOKIE['NODESESSID']) === 'undefined') {		
		var pool = '0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz';
		var newid = '';
		for (var i = 0; i < 26; ++i) {
			var r = Math.floor(Math.random() * pool.length);
			newid += pool.charAt(r);
		}
		pre._COOKIE['NODESESSID'] = newid;
		sessions[pre._COOKIE['NODESESSID']] = {};
		var cookies = 'NODESESSID='+pre._COOKIE['NODESESSID'] + '; Path=/';
		$['res'].setCookie(cookies);
	}
	var id = pre._COOKIE['NODESESSID'];
	if ((typeof sessions[id]) == 'undefined') {
		sessions[id] = {};
	}
	pre._SESSION = sessions[id];
}

exports.setModules = function(moduleObject) {
	$ = moduleObject;
};

exports.GET = function() {
	return pre._GET;
}

exports.POST = function() {	
	return pre._POST;
}

exports.COOKIE = function() {	
	return pre._COOKIE;
}

exports.setcookie = function(name, value, expire, path, secure, httponly) {
	httponly = httponly || false;
	secure = secure || false;
	path = path || '/';
	cookie = name + '=' + value + '; Path=' + path + '; ';
	/*if((typeof domain) != 'undefined') {
		cookie += 'Domain=' + domain + '; ';
	}*/
	if((typeof expire) != 'undefined') {
		cookie += 'Expires=' + expire + '; ';
	}
	if(secure) {
		cookie += 'Secure; ';
	}
	if(httponly) {
		cookie += 'HttpOnly; ';
	}
	console.log(cookie);
	$['res'].setCookie(cookie);
}

exports.REQUEST = function() {	
	return pre._REQUEST;
}

exports.SESSION = function() {	
	return pre._SESSION;
}

exports.setSession = function(name, value) {
	sessions[pre._COOKIE['NODESESSID']][name] = value
	pre._SESSION[name] = value;
}

exports.unsetSession = function(name) {
	delete sessions[pre._COOKIE['NODESESSID']][name];
	delete pre._SESSION[name];
}
